### PR TITLE
Get Getmatches ready for taproot

### PR DESCRIPTION
**Pull request by sageprogrammer**
Created a Draft `GetKeyInformations(OutPoint[] outpoints)` function that gets `KeyPathInforation` for each of the `Outpoint`s provided.
- In future this function will fetch from the DB.

Refactored the code of GetMatches function for taproot

### CODE REVIEW

File: NBXplorer/Repository.cs [hunk 1]
```diff
@@ -1149,67 +1173,154 @@ public Task<TrackedTransaction[]> GetMatches(NBitcoin.Transaction tx, uint256 bl
 		{
 			return GetMatches(new[] { tx }, blockId, now, useCache);
 		}
+
+		private bool isTaproot(TxIn input)
+		{
+			// return true;
+			return input.ScriptSig.IsScriptType(ScriptType.Taproot);
```
Comment (NicolasDorier) [inline NBXplorer/Repository.cs:1180]: The whole problem is that this is not possible. You can't know if a txin is taproot just by looking the input.ScriptSig.

For this reason, I think taproot input/outputs should not be treated differently from the rest.

File: NBXplorer/Repository.cs [hunk 2]
```diff
@@ -809,6 +809,30 @@ private static SavedTransaction ToSavedTransaction(Network network, ReadOnlyMemo
...
+		public async Task<MultiValueDictionary<OutPoint, KeyPathInformation>> GetKeyInformations(OutPoint[] outpoints)
+		{
+			// This is a unoptimized code, written so that it just gives the expected output.
+			// Should be replaced later on by a code like in GetKeyInformations(Script[] scripts)
+			MultiValueDictionary<OutPoint, KeyPathInformation> result = new MultiValueDictionary<OutPoint, KeyPathInformation>();
+			Dictionary<OutPoint, Script> outpointToScript = new Dictionary<OutPoint, Script>();
+			var scripts = new List<Script>();
+			foreach (var outpoint in outpoints)
+			{
+				var prevOutTxn = getTransaction(outpoint.Hash);
```
Comment (NicolasDorier) [inline NBXplorer/Repository.cs:821]: This is sadly not possible, the server would never been able to keep up with the blockchain if we needed 1 roundtrip for each outpoint.

**Comment by NicolasDorier** [comment]
I think in the first PR you should completely disregard taproot, and just make sure to index both `scriptpubkey -> keyinfo` and `outpoint -> keyinfo`. Then trying to fetch both when looking for a match.

To see if it works correctly, you can comment out where you index `scriptpubkey -> keyinfo` for each Txout. Normally tests should still pass.

File: NBXplorer/Repository.cs (hunk 1 above)
Comment (sageprogrammer) [reply NBXplorer/Repository.cs:1180]: Okay will do that.

Comment (sageprogrammer) [reply NBXplorer/Repository.cs:1180]: Okay will do that.

File: NBXplorer/Repository.cs (hunk 2 above)
Comment (sageprogrammer) [reply NBXplorer/Repository.cs:821]: Initially I was also against the idea of making a round trip for each outpoint. But I could not find a function through which I could collectively query all the outpoints for the corresponding txns. This part of code was a sort of place holder for such a function.

Comment (NicolasDorier) [reply NBXplorer/Repository.cs:821]: not only this, but txindex is not mandatory for nbxplorer.

Comment (sageprogrammer) [reply NBXplorer/Repository.cs:821]: What do you mean by txindex not mandatory for nbxplorer?

Comment (sageprogrammer) [reply NBXplorer/Repository.cs:821]: Is it that we do not have to query for every txid we encounter, only the saved ones.

**Pull request closed without merging**